### ISSUE TITLE
Add Data.Hashmap.withEmpty

### ISSUE THREAD

**Issue by bakhtiyarneyman** [tweag/linear-base#118]
**Is your feature request related to a problem? Please describe.**
The only way to create a HashMap is `singleton`, but it's awkward to use because in many settings you don't have the first value to initially populate your HashMap with. E.g. often one needs to make a HashMap out of a list, and the list might be empty.

**Describe the solution you'd like**
Add `withEmpty :: Keyed k => (HashMap k v #-> Unrestricted b) -> Unrestricted b`. `with` in the name of the function is chosen to make it obvious to the reader that `withEmpty` expects a modifier. E.g. `withEmpty (fillFromList xs >>> runAlgorithm)`  as opposed to `empty (fillFromList xs >>> runAlgorithm)`. Similarly `singleton` should be renamed to `withSingleton`.

**Describe alternatives you've considered**
Dropping `singleton` altogether. Rejected because it doesn't seem necessary.

**Additional context**
N/A

**Comment by aspiwack** [comment]
I'm not sure about the `with` thing. I makes sense, but if there isn't a corresponding `empty`. I'm not sure. I will add that to the bikeshedding tickets that I have planed.

Other than that: yes. It was, in fact always planed, and if we don't already have a ticket, it's an oversight.

Not much is going to happen (read: pretty much nothing) over the next two weeks because of people being out of office. Until then, thanks a lot for your interest!

**Comment by Divesh-Otwani** [comment]
Hi @bakhtiyarneyman  -- I think this is important and I'll aim to get a PR in today for this.

(Also, for completeness I'm noting this is a case of #100 )

**Comment by Divesh-Otwani** [comment]
Hi @bakhtiyarneyman See #122.

**Comment by Divesh-Otwani** [comment]
I'm rubber ducking a bit to help move things along since this is getting stranger and stranger:

* In the implemenation below, the test suite fails and the function that probes (`queryIndex`) runs forever. Printing it out, it seems like the array gets full and the resizing function does not work. 

```
-- | Run a computation using an empty hashmap
empty :: (Keyed k, Default k, Default v) =>
  (HashMap k v #-> Unrestricted b) -> Unrestricted b
empty (f :: HashMap k v #-> Unrestricted b) =
  alloc defaultSize (error "Empty value") applyHMfunc hmap
  where
    applyHMfunc :: Array (k, v, PSL) #-> Unrestricted b
    applyHMfunc arr = f $ HashMap (defaultSize, 0) arr
```

* By adding a bang in the last line forcing the evaluation of the array, the test suite passes:
(Also, this took a while to discover.)

```
-- | Run a computation using an empty hashmap
empty :: (Keyed k, Default k, Default v) =>
  (HashMap k v #-> Unrestricted b) -> Unrestricted b
empty (f :: HashMap k v #-> Unrestricted b) =
  alloc defaultSize (error "Empty value") applyHMfunc hmap
  where
    applyHMfunc :: Array (k, v, PSL) #-> Unrestricted b
    applyHMfunc !arr = f $ HashMap (defaultSize, 0) arr
```

* I tried to see if I could recreate a failing test using GHCi and an older version of GHC. I could not due to stack overflows on even simple things:

```
test :: HashMap Int String #-> Unrestricted Bool
test hmap = doNothing hmap

doNothing :: HashMap Int String #-> Unrestricted Bool
doNothing !hmap = lseq hmap (Unrestricted True)

```

These persist on analogous calls to linear mutable arrays.

* The stack overflows seem to occur when you load the file but not when you write in GHCi directly. I'm currently confused as to why this is the case.

**Comment by Divesh-Otwani** [comment]
This issue is now solved in PR #124 -- sorry for the delays and complications @bakhtiyarneyman. This helped reveal some deep and complex bugs :)
